### PR TITLE
Add type hints to `synapse/storage/databases/main/room.py`

### PR DESCRIPTION
**Pull request by squahtx**
* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -1020,7 +1020,7 @@ async def transfer_room_state_on_room_upgrade(
         # Add new room to the room directory if the old room was there
         # Remove old room from the room directory
         old_room = await self.store.get_room(old_room_id)
-        if old_room and old_room["is_public"]:
+        if old_room is not None and old_room["is_public"]:
```
Comment (squahtx) [inline synapse/handlers/room_member.py:1023]: Updated for consistency with the argument to `add_room_to_group` below.

File: synapse/storage/databases/main/__init__.py [hunk 2]
```diff
@@ -149,7 +149,6 @@ def __init__(
             ],
         )
 
-        self._event_reports_id_gen = IdGenerator(db_conn, "event_reports", "id")
```
Comment (squahtx) [inline synapse/storage/databases/main/__init__.py:152]: `_event_reports_id_gen` is only used in `RoomStore`

File: synapse/storage/databases/main/room.py [hunk 3]
```diff
@@ -1448,7 +1472,7 @@ def _background_populate_rooms_creator_column_txn(txn: LoggingTransaction):
         return batch_size
 
 
-class RoomStore(RoomBackgroundUpdateStore, RoomWorkerStore, SearchStore):
+class RoomStore(RoomBackgroundUpdateStore, RoomWorkerStore):
```
Comment (squahtx) [inline synapse/storage/databases/main/room.py:1475]: `SearchStore` isn't needed as a parent class and makes Python fail to come up with an mro.

**Review by DMRobertson — APPROVED** [review]
Thanks for confirming. I'm happy with this!

**Pull request merged**